Update MicroProfile Health guide to MicroProfile Health 2.0 (#3129)
quarkusio/quarkus: docs/src/main/asciidoc/health-guide.adoc

== Running the health check

Importing the `smallrye-health` extension directly exposes three REST endpoints:

- `/health/live` - The application is up and running.

> - `/health/ready` - The application is ready to serve requests.

Reviewer: I would put this one first.

> - `/health` - Accumulating all health check procedures in the application.

Reviewer: Someone told me it was deprecated? Should we talk about it?

Author: The qualifier is, not the endpoint, at least not in 2.0 up to my knowledge.

> To check that smallrye-health extension is working as expected:

Reviewer: ```suggestion
To check that the `smallrye-health` extension is working as expected:
```

* `checks` -- an array of individual checks

The general `status` of the health check is computed as a logical AND of all the 
declared health check procedures. The `checks` array is empty as we have not specified 
any health check procedure yet so let's define some.

== Creating your first health check

In this section, we create our first simple health check procedure.

Create the `org.acme.health.SimpleHealthCheck` class:

[source,java]
----
package org.acme.health;

import org.eclipse.microprofile.health.HealthCheck;
import org.eclipse.microprofile.health.HealthCheckResponse;
import org.eclipse.microprofile.health.Liveness;

import javax.enterprise.context.ApplicationScoped;

@Liveness
@ApplicationScoped
public class SimpleHealthCheck implements HealthCheck {

    @Override
    public HealthCheckResponse call() {
        return HealthCheckResponse.named("Simple health check").up().build();
    }
}
----

As you can see health check procedures are defined as implementations of the 
`HealthCheck` interface which are defined as CDI beans with the one of the 
following CDI qualifiers: 

- `@Liveness` - the liveness check accessible at `/health/live`

> - `@Readiness` - the readiness check accessible at `/health/ready`

Reviewer: Same here, I would put this one first.

> - `@Health` - *deprecated* unspecified health procedure accessible at `/health`

Reviewer: I would remove this one: Quarkus is not 1.0 yet so let's not advertise deprecated features.

WARNING: The `@Health` qualifier is deprecated in version 2.0 of MicroProfile 
Health specification and it can be removed in any later release.

NOTE: The `/health` endpoint accumulates all `@Liveness`, `@Readiness` and deprecated 
`@Health` procedures and it is possible it will be removed in a later release of the 

> specification.  

Reviewer: I would remove these 2 notes and just advertise how things should be done now.

`HealthCheck` is a functional interface whose single method `call` returns a 
`HealthCheckResponse` object which can be easily constructed by the fluent builder 
API shown in the example.

As we have started our Quarkus application in dev mode simply repeat the request 
to `http://localhost:8080/health/live` by refreshing your browser window or by 
using `curl http://localhost:8080/health/live`. Because we defined our health check 
to be a liveness procedure (with `@Liveness` qualifier) the new health check procedure 
is now present in the `checks` array.

Congratulations! You've created your first Quarkus health check procedure. Let's 
continue by exploring what else can be done with the MicroProfile Health specification.

== Adding a readiness health check procedure

In the previous section, we created a simple liveness health check procedure which states 
whether our application is running or not. In this section, we will create a readiness 
health check which will be able to state whether our application is able to process 

> requests.

Reviewer: I must admit I would talk about readiness first as it seems more logical.

But maybe there's a good reason for this order?

We will create another health check procedure that simulates a connection to 
an external service provider such as a database. For starters, we will always return 
the response indicating the application is ready.

Create `org.acme.health.DatabaseConnectionHealthCheck` class:

[source,java]
----
package org.acme.health;

import org.eclipse.microprofile.health.HealthCheck;
import org.eclipse.microprofile.health.HealthCheckResponse;
import org.eclipse.microprofile.health.Readiness;

import javax.enterprise.context.ApplicationScoped;

@Readiness
@ApplicationScoped
public class DatabaseConnectionHealthCheck implements HealthCheck {

    @Override
    public HealthCheckResponse call() {
        return HealthCheckResponse.named("Database connection health check").up().build();
    }
}

----

If you now rerun the health check at `http://localhost:8080/health/live` the `checks` 

> array will contain only previously defined `SimpleHealthCheck` as it is the only 

Reviewer: ```suggestion
array will contain only the previously defined `SimpleHealthCheck` as it is the only 
```

> defined check with the `@Liveness` qualifier. However, if you access 

Reviewer: ```suggestion
check defined with the `@Liveness` qualifier. However, if you access 
```

> `http://localhost:8080/health/ready` (in the browser or by 

Reviewer: ```suggestion
`http://localhost:8080/health/ready` (in the browser or with 
```

`curl http://localhost:8080/health/ready`) you will see only the 
`Database connection health check` as it is the only health check defined with the 
`@Readiness` qualifier as the readiness health check procedure.

> NOTE: If you access `http://localhost:8080/health` you will get back both checks.

Reviewer: Same here, I would remove that.

Author: `/health` is still present in MP Health 2.0. There is a possibility that it will be extended in the future. I don't think it's deprecated so I would keep the note for now.

> More information about which health check procedures should be used when is detailed 

Reviewer: ```suggestion
More information about which health check procedures should be used in which situation is detailed 
```

in the MicroProfile Health specification. Generally, the liveness procedures determine 
whether the application should be restarted while readiness procedures determine 
whether it makes sense to contact the application with requests.

== Negative health check procedures

In this section, we extend our `Database connection health check` with the option of 

> stating that our application is not ready to process requests as underlying database 

Reviewer: ```suggestion
stating that our application is not ready to process requests as the underlying database 
```

}
----

If you now rerun the readiness health check (at `http://localhost:8080/health/ready`) 
the overall `status` should be DOWN. You can also check the liveness check at 
`http://localhost:8080/health/live` which will return the overall `status` UP because 
it isn't influenced by the readiness checks.

As we shouldn't leave this application with a readiness check in a DOWN state and 

> because we are running Quarkus dev mode you can add `database.up=true` in 

Reviewer: ```suggestion
because we are running Quarkus in dev mode you can add `database.up=true` in 
```

`src/main/resources/application.properties` and rerun the readiness health check again 
-- it should be up again.

== Adding user-specific data to the health check response

In previous sections, we saw how to create simple health checks with only the minimal 
attributes, namely, the health check name and its status (UP or DOWN). However, the 
MicroProfile specification also provides a way for the applications to supply 
arbitrary data in the form of key-value pairs sent to the consuming end. This can be 
done by using the `withData(key, value)` method of the health check response 
builder API.

> Let's create new health check procedure `org.acme.health.DataHealthCheck`:

Reviewer: ```suggestion
Let's create a new health check procedure `org.acme.health.DataHealthCheck`:
```
